### ISSUE TITLE
DataTable: Removed data (rows) remain after update when filtered before

### ISSUE THREAD

**Issue by MatthiasPischka** [primefaces/primefaces#7336]
**Describe the defect**
When using a List (for example ArrayList) of DTOs for a dataTable and removing some elements from the list, updating the table throw ajax, everythink works fine.
But when I first filter the rows inside the dataTable throw column or global Filter and then remove some of the still visible elements, they remain after ajax update. But the elements are arleady removed from the Java List.

The Bug was also discussed in the forum: https://forum.primefaces.org/viewtopic.php?f=3&t=66918

**Reproducer**
Checkout https://github.com/MatthiasPischka/primefaces-test.git and run as described (mvn jetty:run for example).
After starting, visit the page http://localhost:8080/primefaces-test/data-table-remove-bug.xhtml (you get there throw navigation from default page, too).
Working and buggy example are described on the page.

**Environment:**
 - PF Version: _10.0.0_
 - JSF + version: _Mojarra 2.3.14.SP01_
 - Affected browsers: _Testet in Firefox and Chrome_

**Comment by melloware** [comment]
I have confirmed this issue.  Even calling on the server side Java code `this.filteredList.remove(product);` is not removing the product from the filtered view.

**Comment by christophs78** [comment]
We had similar issues in one of our (still PF 7 based) applications too. In our case when adding new rows or editing existing ones on an filtered DataTable.
From my feeling we still have not worked out all issues caused by having the full list an the filtered list.
@Rapster : Some thoughts?

**Comment by christophs78** [comment]
Propably we should write integration-tests for at least 3 usecases for filtered DataTable:

- add new object to list
- remove object from list
- edit existing object (and afterwards do not fullfill filter-criterias)

**Comment by melloware** [comment]
I use lazy tables almost exclusively and I do not remember seeing this issue. But without lazy this is definitely happening.

**Comment by christophs78** [comment]
@MatthiasPischka: Can you try following code?
```java
public void remove(Product product) {
	System.out.println("Product to remove: " + product.getLabel());
	this.productList.remove(product);
//    this.filteredList.remove(product);

	DataTable dataTable = (DataTable)FacesContext.getCurrentInstance().getViewRoot().findComponent("form:dt-products");
	FilterFeature filterFeature = (FilterFeature) dataTable.getFeature(DataTableFeatureKey.FILTER);
	filterFeature.filter(FacesContext.getCurrentInstance(), dataTable);
	SortFeature sortFeature = (SortFeature) dataTable.getFeature(DataTableFeatureKey.SORT);
	sortFeature.sort(FacesContext.getCurrentInstance(), dataTable);
}
```

**Comment by christophs78** [comment]
This issue is even reproduceable with the CRUD-example of premium templates: eg https://www.primefaces.org/freya/crud.xhtml
At this point it´s hard to argue against people who tell you "yes, the showcase looks good, but in real world there are way to many pitfalls when you start combining featues".

**Comment by MatthiasPischka** [comment]
> 
> 
> @MatthiasPischka: Can you try following code?
> 
> ```java
> public void remove(Product product) {
> 	System.out.println("Product to remove: " + product.getLabel());
> 	this.productList.remove(product);
> //    this.filteredList.remove(product);
> 
> 	DataTable dataTable = (DataTable)FacesContext.getCurrentInstance().getViewRoot().findComponent("form:dt-products");
> 	FilterFeature filterFeature = (FilterFeature) dataTable.getFeature(DataTableFeatureKey.FILTER);
> 	filterFeature.filter(FacesContext.getCurrentInstance(), dataTable);
> 	SortFeature sortFeature = (SortFeature) dataTable.getFeature(DataTableFeatureKey.SORT);
> 	sortFeature.sort(FacesContext.getCurrentInstance(), dataTable);
> }
> ```

@christophs78 Yes this workaround brings the expected behaviour (product is removed even when filtered before). I added a second button to my example (https://github.com/MatthiasPischka/primefaces-test).
And yes you are right. I found this issue while i am using the avalon theme and orienting at the crud example.

**Comment by christophs78** [comment]
Think we should try to fix this. (And not only add the convenience function suggested by my PR.)

`value`/`filteredValue` after `FilterFeature#filter` (2 values filtered)
![image](https://user-images.githubusercontent.com/10461942/119540329-9ef7a400-bd8d-11eb-84d1-0a5703685e7b.png)
value is (internally) set to the filteredValue

One request-cycle later it looks like this:
![image](https://user-images.githubusercontent.com/10461942/119540454-c0f12680-bd8d-11eb-9516-c3d1fc904f5e.png)

`DataTableRenderer#encodeTbody` always renders `value` (not `filteredValued`)
So adding the new row to filteredValue (or removing it from filteredValue does not work anymore)
![image](https://user-images.githubusercontent.com/10461942/119539204-3825bb00-bd8c-11eb-948f-dcbee827ba65.png)

Good question how to solve this correct and not do another hack which may fix this one and creates some other issue(s).

**Comment by melloware** [comment]
Agreed that does seem wonky and don't want to add another hack on top of it.

**Comment by christophs78** [comment]
Cross-checked this with PF 8.0. One request-cycle after filter is applied value and filteredValue points to the same ArrayList-instance. (PF 10.0.0 has at this point two different ArrayList-instances as shown before.)

**Comment by christophs78** [comment]
PF 10.0.0 - saved state for DataTable with filter
![image](https://user-images.githubusercontent.com/10461942/119567206-53a0be00-bdac-11eb-8781-8e0387686d55.png)

PF 8.0 - saved state for DataTable with filter (exact same usecase)
![image](https://user-images.githubusercontent.com/10461942/119567259-661af780-bdac-11eb-8c36-48fb0a467eb0.png)

IMO this shows PF 8.0 does not save state for value-attribute. (--> so we get back the same instance on the next request cylce)
PF 10.0.0 saves state for value-attribute. (--> so we get back a new instance on the next requests cycle causing trouble in this usecase)

@Rapster - your opinion would be appreciated

**Comment by Rapster** [comment]
I'll try to have a look this week

**Comment by tandraschko** [comment]
maybe bit related to: https://github.com/primefaces/primefaces/issues/1390

**Comment by tandraschko** [comment]
i remember that we had many users with errors, because their POJO doesnt implement Serializable
i think it should not be saved in the state, yes
just always recevied from the bean

not sure why it happens

**Comment by tandraschko** [comment]
yep, reproduced
POJO without Serialization works fine in 8.0, thats really that value seems no saved in component state

**Comment by tandraschko** [comment]
ok, in 8.0 the DataTable#setValue is called in the filter request with:
1) null
2) List with filtered POJOs
3) null

in 10.0:
1) null
2) List with filtered POJOs



it seems that this is missing in 10.0:

```
    @Override
    public Object saveState(FacesContext context) {
        if (isFilteringEnabled()) {
            setValue(null);
        }
```

**Comment by tandraschko** [comment]
so we have 2 possibilites:
1) use either value or filteredValues in the DataTableRenderer
2) reset like we did in 8.0

**Comment by tandraschko** [comment]
reintroduced that statement
could @melloware or @christophs78 create automatic tests for this?
one example is here, the other one is in 7367, when the POJO doesnt have the Serializable

**Comment by melloware** [comment]
Let me kick off the current tests.  And then we need to add more tests for this.

**Comment by melloware** [comment]
Multiple integration tests fail now...
```
 Failures: 
Error:    DataTable001Test.testFilter:155->AbstractDataTableTest.assertRows:74->AbstractDataTableTest.assertRows:83 expected: <2> but was: <3>
Error:    DataTable007Test.testFilterAndEditRow_1442:151 expected: <abc> but was: <JavaScript>
Error:    DataTable009Test.testFilterIssue7026:167 expected: <2020> but was: <1990>
Error:    DataTable010Test.testSelectionMultipleWithPagingAndFilteringAndDelete:127->testSelectionMultipleWithPagingAndFiltering:119 expected: <1,3,5> but was: <1,2,5>
Error:    DataTable010Test.testSelectionMultipleWithPagingAndFiltering:119 expected: <1,3,5> but was: <1,2,5>
Error:    DataTable018Test.testListModelFilter:152->AbstractDataTableTest.assertRows:74->AbstractDataTableTest.assertRows:83 expected: <2> but was: <3>
Error:  Errors: 
Error:    DataTable009Test.testFilterIssue1390:66 » NoSuchElement no such element: Unabl...
Error:    DataTable009Test.testKeepSelectOneMenuFilter:114->validateFirstAppeared:174 » NoSuchElement
```

**Comment by tandraschko** [comment]
hmm, not sure why that happens now

Im out for today, "feierabend" in germany now.

**Comment by christophs78** [comment]
We already have a PR with additional tests for this: https://github.com/primefaces-extensions/primefaces-integration-tests/pull/156

**Comment by tandraschko** [comment]
@melloware i copied the DataTable001Test to a primefaces-test but i works fine for me? can you verify?

**Comment by tandraschko** [comment]
ok, found it
monkeypatched the DataTable in the test-project and forgot to delete it, sorry

should be better now, can you trigger the tests again?

**Comment by christophs78** [comment]
```
Error:  Failures: 
Error:    DataTable008Test.testFilterIssue_5481:77->AbstractDataTableTest.assertRows:74->AbstractDataTableTest.assertRows:83 expected: <2> but was: <1>
[INFO] 
Error:  Tests run: 309, Failures: 1, Errors: 0, Skipped: 0
```

You can rerun the job on your own via
![image](https://user-images.githubusercontent.com/10461942/119946969-7cd17200-bf97-11eb-8b47-83d50a061e9d.png)

**Comment by tandraschko** [comment]
bah
i basically restored #5481 and i think the behavior was partly correct (they were restored on postback only, not on pageload/f5)
the filteredValues are stored in the SessionScoped and therefore it should be restored on f5 actually, or not?
the problem is that the filter-input doesnt contain the old text ("2010" in the test) anymore as its component state (viewscoped)

so one state is stored in sessionscoped bean, the other state is stored in viewscoped component state, thats a problem

we have 3 possible solutions
1)  should we say, filteredValues should not be used with SessionScoped?
2) make filteredValues write-only but e.g. my customer uses it in read-only manner
3) add a note to the docs, that when used SessionScoped, you also have to maintain filterBy and save it in session (is this possible?) So that filteredValues and filterBy matches when pressing f5

**Comment by christophs78** [comment]
https://github.com/primefaces-extensions/primefaces-integration-tests/blob/master/src/test/java/org/primefaces/extensions/integrationtests/datatable/DataTable008Test.java#L45 Is broken now
DataTable008Test.testFilterIssue_5481:77->AbstractDataTableTest.assertRows:74->AbstractDataTableTest.assertRows:83 expected: <2> but was: <1>

**Comment by tandraschko** [comment]
Will continue the broken case in a new ticket

**Comment by christophs78** [comment]
@tandraschko Maybe we can in future try to do such fixes via PR´s which we squash to one commit. This would make it way easier for PrimeTek to pick such fixes for a elite-release without  the risk of breaking something because they only select two of three commits or whatever.
Or someone else is confused at a later point when looking into the commit history of a class.

**Comment by buunsal** [comment]
> @tandraschko Maybe we can in future try to do such fixes via PR´s which we squash to one commit. This would make it way easier for PrimeTek to pick such fixes for a elite-release without the risk of breaking something because they only select two of three commits or whatever.
> Or someone else is confused at a later point when looking into the commit history of a class.

Absolutely I agree. @tandraschko I want to fix this issue for 10.0.5 but i don't understand too much. So I can't add it because I don't want to break anything. Could you please help with what I should include?

**Comment by doteo78** [comment]
I tried it with version 10.0.6 and the problem still persists. After using the filter the datatable is not updated (new rows/removed rows) via ajax action. You need to use the global filter again to see the changes.
Could you check it please?

**Comment by buunsal** [comment]
@doteo78 I already tested before the version was released. Could you please create a reproducer?

**Comment by christophs78** [comment]
We added at least one integrationtest for this. You may run this one against
10.0.6. (I have to look which one.)

**Comment by christophs78** [comment]
DataTable021Test should check this.
@doteo78 does it work for you with 11.0.0-SNAPSHOT?

**Comment by buunsal** [comment]
@christophs78 how can i run this test ?

**Comment by christophs78** [comment]
https://github.com/primefaces/primefaces/tree/master/primefaces-integration-tests

When you use IntelliJ it should be easy once you downloaded Chromedriver. I may come back later this day with more detailed information. (Currently on vacation and out in the nature.)

**Comment by buunsal** [comment]
https://github.com/primefaces/primefaces/tree/master/primefaces-integration-tests this project is not compiling with 10.0.6. There are some differences in version 11.0.0 and 10.0.6 for example tree datatable class structure etc. So at least I couldn't get it to work 😅  I would love to know when you explain it in more detail.

**Comment by doteo78** [comment]
> DataTable021Test should check this.
> @doteo78 does it work for you with 11.0.0-SNAPSHOT?

Tested with master-SNAPSHOT version and it does not work for me :-(

**Comment by christophs78** [comment]
@sqores: https://github.com/christophs78/primefaces/tree/IT_PF_10_0_X let´s you run our current integrationtests against 10.0.x.

@doteo78: So there must be another issue not solved yet. --> We need as @sqores mentioned at least a reproducer or even better a PR containing a failing integrationtest.

**Comment by melloware** [comment]
Yes @doteo78 your use case must be different than the others so if you can provide a [PrimeFaces Test](https://github.com/primefaces/primefaces-test)  reproducer we can debug it.

**Comment by doteo78** [comment]
[primefaces-test-7336.zip](https://github.com/primefaces/primefaces/files/7014714/primefaces-test-7336.zip)

Run the attached project (mvn jetty:run) and visit http://localhost:8080/primefaces-test/

Click "Delete row" button -> The first row is deleted and the datatable is updated. Now if you use the filter and leave it blank again you see the rows, but if you click "Delete row" the row is deleted but the datatable is not updated until using the filter again.

Hope it helps!

**Comment by melloware** [comment]
I can confirm your issue exists on 11.0.0-SNAPSHOT and have re-opened the ticket.

**Comment by christophs78** [comment]
I´m writing an (currently failing) integrationtest for this. @doteo78 : Thanks for providing the reproducer!

**Comment by christophs78** [comment]
@doteo78 : After some research and trying to writing a IT i think to know what´s going on.
DataTable requires you to remove the row from the filtered list too. See our already existing integrationtest: https://github.com/primefaces/primefaces/blob/17d7bc92d52acab43ab6a21c648f93391df0ded0/primefaces-integration-tests/src/main/java/org/primefaces/integrationtests/datatable/DataTable021.java#L66

PrimeFaces 11 adds a new method to DataTable to help you with this:

```java
public void deleteProgrammingLanguage() {
    ProgrammingLanguage java = this.progLanguages.stream().filter(l -> l.getName().equals("Java")).findFirst().get();
    this.progLanguages.remove(java);

    DataTable dataTable = (DataTable) FacesContext.getCurrentInstance().getViewRoot().findComponent("form:datatable");
    dataTable.filterAndSort();
}
```

@tandraschko , @Rapster , @melloware : Maybe we should add a note regarding this to Showcase and DataTable-docu? Whoever reads this. ;-)

**Comment by doteo78** [comment]
@christophs78 - In the beginning of the ticket while debugging the initial issue you mentioned this:
> So adding the new row to filteredValue (or removing it from filteredValue does not work anymore)

Besides, when adding a new row it is more complicated to know if it is neccesary to add the object to filteredValue. You need to check first if it meets the filter criteria.

So for now, seeing what you guys are doing in Primefaces 11 I am calling this function after adding or removing (have not tried editing yet) a row:

```java
private void filterAndSort() {
	DataTable dataTable = (DataTable) FacesContext.getCurrentInstance().getViewRoot()
			.findComponent("frmTest:singleDT");
	FilterFeature filterFeature = (FilterFeature) dataTable.getFeature(DataTableFeatureKey.FILTER);
	filterFeature.filter(FacesContext.getCurrentInstance(), dataTable);
	SortFeature sortFeature = (SortFeature) dataTable.getFeature(DataTableFeatureKey.SORT);
	sortFeature.sort(FacesContext.getCurrentInstance(), dataTable);
}
```

What do you think?

**Comment by christophs78** [comment]
> @christophs78 - In the beginning of the ticket while debugging the initial issue you mentioned this:
> 
> > So adding the new row to filteredValue (or removing it from filteredValue does not work anymore)

This was true at the time of writing this. This was fixed.

Your solution makes sense.

DataTable overall is way more complicated as most of us (including me) think first. Think eg of combining filter, paging and editable. You may add a new row on a page currently not selected. Or adding a (new/blank) row which does not fit into filter-criterias.
At some point i accepted it´s hard (almost impossible) for DataTable to do always automatically "the right thing". So there are situations where the developer using DataTable has to do "some things" to achieve what he wants.

**Comment by doteo78** [comment]
[primefaces-test-7336-v2.zip](https://github.com/primefaces/primefaces/files/7046723/primefaces-test-7336-v2.zip)

Another bug only when using MyFaces (a reproducer is attached).

Run the project (mvn jetty:run -Pmyfaces23) and visit http://localhost:8080/primefaces-test/
Then click "Delete row" on any row -> the row is deleted and the datatable is updated.
Click again "Delete row" on another row -> the row is deleted but the datatable is NOT updated.

It only happens when the button is on the row and after calling filterAndSort() method. If you click "Delete first row" or use the filter the datatable is updated properly.

Tested with version 10.0.6. I suspect it won't work with 11.0.0-SNAPSHOT either but I could not run the reproducer with this version due to an OpenWebBeans exception.

**Comment by melloware** [comment]
Here is the MyFaces 23 reproducer using 11.0.0-SNAPSHOT and you are correct the second delete button does not update the table.

[pf-7336-myfaces23.zip](https://github.com/primefaces/primefaces/files/7046888/pf-7336-myfaces23.zip)

cc @tandraschko

**Comment by tandraschko** [comment]
is this the same: https://github.com/primefaces/primefaces/actions/runs/1165377212

@christophs78 can you check?

**Comment by christophs78** [comment]
> is this the same: https://github.com/primefaces/primefaces/actions/runs/1165377212
> 
> @christophs78 can you check?

Problably not the same. If it would be the same it will fail for Java 8 too.
Currently looking into what is going on here.

**Comment by christophs78** [comment]
@doteo78 : Please try to add `if (dataTable.getFilteredValue()!=null) {` to your filterAndSort-method.

**Comment by christophs78** [comment]
Looks like a have to accept DataTable#filterAndSort does not work with MyFaces. (@tandraschko : Maybe you can add some aspects to this.)
When we don´t get it working with MyFaces we will have to remove it.

@doteo78 : Plan B is to do it this way. Well knowing this is not ideal and has some other limitations.
```java
public void removeV1(ProgrammingLanguage programmingLanguage) {
	this.progLanguages.remove(programmingLanguage);
	if (this.filteredProgLanguages != null) {
		this.filteredProgLanguages.remove(programmingLanguage); // see https://github.com/primefaces/primefaces/issues/7336
	}
}
```

**Comment by doteo78** [comment]
> @doteo78 : Please try to add `if (dataTable.getFilteredValue()!=null) {` to your filterAndSort-method.

It does not work when using the filter first as you probably already know.

> @doteo78 : Plan B is to do it this way. Well knowing this is not ideal and has some other limitations.

Yep! This is what I am doing right now ;-)

Another thing I realized is that search keywords are lost after updating the datatable (adding/removing rows) although the datatable is still well filtered. It seems to be fixed in version 11.0.0-SNAPSHOT and I don't know if it is related with issue #7325. Behaviour is similar but I am not using MultiViewState.

Any chance to fix this in an elite 10 version?
Do you prefer to discuss this in a separate issue?

**Comment by DazzlingBob** [comment]
Just checking. Is it correct that after this patch (primefaces version => 10.0.6):
When adding or removing list items you no longer need to add/remove them from the filteredList?
I used to do this in primefaces 8.

And what about inline editing of an item? No clear and addAll of the filteredList needed?
